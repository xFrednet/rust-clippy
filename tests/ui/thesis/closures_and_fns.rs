//@rustc-env: CLIPPY_PETS_PRINT=1
//@rustc-env: CLIPPY_PRINT_MIR=1

#[derive(Default)]
struct Example {
    owned_1: String,
    owned_2: String,
    copy_1: u32,
    copy_2: u32,
}

#[warn(clippy::borrow_pats)]
fn use_local_func() {
    let func = take_string_ref;

    func(&String::new());
}

#[warn(clippy::borrow_pats)]
fn use_arg_func(func: fn(&String) -> &str) {
    func(&String::new());
}

#[warn(clippy::borrow_pats)]
fn call_closure_with_arg(s: String) {
    let close = |s: &String| s.len();
    close(&s);
}

#[warn(clippy::borrow_pats)]
fn call_closure_borrow_env(s: String) {
    let close = || s.len();
    close();
}

#[warn(clippy::borrow_pats)]
fn call_closure_move_s(s: String) {
    let close = move || s.len();
    close();
}

#[warn(clippy::borrow_pats)]
fn call_closure_move_field(ex: Example) {
    let close = move || ex.owned_1.len();
    close();
}

fn take_string(_s: String) {}
fn take_string_ref(_s: &String) {}
fn pass_t<T>(tee: T) -> T {
    tee
}

fn main() {}
