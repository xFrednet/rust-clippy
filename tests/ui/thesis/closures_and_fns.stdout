# "use_local_func"
- func           : (Immutable, Owned    , Local   , Copy   , NonDrop , Fn       ) {}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {NoArguments}

# "use_arg_func"
- func           : (Immutable, Owned    , Argument, Copy   , NonDrop , Fn       ) {}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {}

# "call_closure_with_arg"
- s              : (Immutable, Owned    , Argument, NonCopy, PartDrop, UserDef  ) {Borrow, CtorBorrow}
- close          : (Immutable, Owned    , Local   , Copy   , NonDrop , Closure  ) {Borrow, ArgBorrow}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {HasTempBorrow}

# "call_closure_borrow_env"
- s              : (Immutable, Owned    , Argument, NonCopy, PartDrop, UserDef  ) {Borrow, ClosureBorrow}
- close          : (Immutable, Owned    , Local   , Copy   , NonDrop , Closure  ) {Borrow, ArgBorrow}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {HasTempBorrow}

# "call_closure_move_s"
- s              : (Immutable, Owned    , Argument, NonCopy, PartDrop, UserDef  ) {Moved, MovedToClosure}
- close          : (Immutable, Owned    , Local   , NonCopy, PartDrop, Closure  ) {Borrow, ArgBorrow}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {HasTempBorrow}

# "call_closure_move_field"
- ex             : (Immutable, Owned    , Argument, NonCopy, PartDrop, UserDef  ) {PartMoved, PartMovedToClosure}
- close          : (Immutable, Owned    , Local   , NonCopy, PartDrop, Closure  ) {Borrow, ArgBorrow}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {HasTempBorrow}

